### DISCUSSION TITLE
Add more i18n configuration options

### DISCUSSION THREAD

**Discussion by maapteh** [vercel/next.js#28004] [Help] [answered]
### Describe the feature you'd like to request

I would like to request for three features in the new feature of i18n setup available since NextJS 10 (awesome start btw!).

## 1. have localeDetection available per domain

The current implementation sees translations not more then a simple transformation of the website. So if i turn it on it is nice that when having english as default in my browser it goes to domain.com instead of domain.nl

But for larger companies domains are more like a Point Of Sale with own rules and an own market. Now the feature request is to have `localeDetection` as option per domain. That will fix the fact for belgium/chinese etc we can just order the languages available on that domain, so not outside. And prevent them to go to a total other domain.

So:
```
      {
        domain: 'mpth.fr',
        locales: ['fr-FR', 'en-FR'],
        defaultLocale: 'fr-FR',
        localeDetection: true,
      },
 ```

## 2. have defaultLocale as optional

For some domains defaults are perfect. So for french website, french makes more sence when also giving english as option. French people like it, certainly when using localeDetection for that domain. But their neighbours will not like that so much. So on Belgium if i make Dutch or French as default, half of the nation will not like that its the default, also SEO etc. When we make it optional, root will simply go to one of the two. Certainly with feature request 1 that will fit perfectly.

```
      {
        domain: 'mpth.be',
        locales: ['nl-BE', 'fr-BE'],
        localeDetection: true,
      },
 ```
 
 ## 3. have locale rewrites as option
 
To have more flexibilty and not having rewrite schema's all over the place maybe its nice to offer locale rewrites. So  allthough using 'nl-BE', there is an option to have it in url like 'nl' by making it transformable. Or even 'nl/be' like Nike/etc. It will give us so much flexibilty.
 
 ```
      {
        domain: 'mpth.be',
        locales: ['nl-BE', 'fr-BE'],
        localeDetection: true,
        rewrites: {
          'nl-BE': 'nl',
          'fr-BE': 'fr'
        }
      },
 ```
 
 or
 
  ```
      {
        domain: 'mpth.com',
        locales: ['nl-BE', 'fr-BE', 'en', etc...],
        localeDetection: true,
        rewrites: {
          'nl-BE': 'nl/be',
          'fr-BE': 'fr/be',
          etc...
        }
      },
 ```

__I know with 3 feature requests in 1 my chances are not so high. But i just want to have mentioned them so i do get a chance to use NextJS in some really bigger domains. Thanks for reading!__

### Describe the solution you'd like

I gave clear examples in my Request. If not clear i would be happy to discuss.

### Describe alternatives you've considered

Writing my own Middleware which hooks up on the internals of NextJS. Which is a path I can't take because it feels dirty and it is.

Deploying application per POS. So each having a different routing setup, label sheet generation (so will be able to use 'nl' etc), and configurable 'locale' stuff on multiple places to make most code re-usable. Now containers cant be re-used when traffic on other domains need more containers, so have to spin them up instead of using the quiet ones (cheaper). But it also doesn't solve issue 2.

Trying to explain clients there is a trade-off in using the most versatile/flexible framework. But these discussions don't go well since in the end we have to deliver what they want.

**Comment by maapteh** [comment]
There was some traction about the optional defaultLocale at https://github.com/vercel/next.js/discussions/18419 but i created the feature request since i really want the combination of the three topics mentioned.

Also the last topic was slightly mentioned at https://github.com/vercel/next.js/discussions/17078#discussioncomment-357232 where they gave Uber as example (we used Nike as example)

I hoped it can become a feature request soon again :) I'm more then willing with helping as long as some experts about internals can help.

**Comment by maapteh** [comment]
It seems im almost there using the NextJS 12 new middleware option, so not using locale settings at all :).

~~Only have a bug in useRouter which uses the internal pathname, not the client url one when proceeding client-side to the next pages.~~

Issue created here https://github.com/vercel/next.js/issues/32313

And solved/bypassed the issue for the useRouter. So basically im not using the nextjs i18n for routings anymore but simple custom middleware. And all 3 options can now be implemented. Thanks for version 12!

**Comment by leerob** [comment, marked as answer]
**Update:** As you mentioned, this is now enabled through [Middleware](https://nextjs.org/docs/middleware) which gives you full flexibility to control your i18n routing.

**Reply by leerob** [reply]
Here's an example for adding a prefix to the default locale: https://github.com/vercel/next.js/discussions/18419#discussioncomment-1561577
